### ISSUE TITLE
statements() and astNodes() steps don't interact well

### ISSUE THREAD

**Issue by vlad902** [octopus-platform/joern#144]
The `statements` and `astNodes` steps don't interact well together. I think it's because they both use repeat steps, but I haven't been able to find anything about how multiple repeat steps in a single query interact.

Here's an example, `node.statements().astNodes()` fails, thought `statement_node.astNodes()` and 
`node.statements().identity().astNodes()` succeeds:
```
> g.V(411545664).statements().astNodes()

> statement = g.V(411545664).statements()
v[411541568]
> g.V(411541568).astNodes()
v[411541568]
v[411820032]
...
> g.V(411545664).statements().identity().astNodes()
v[411541568]
v[411820032]
...
```

`identity` is not the only step that changes this behavior, `barrier`, `unfold`, etc. do as well.

**Comment by vlad902** [comment]
@timhemel Maybe this is related to b5485e4a40a83485906970d033701874faeaa122 ?

**Comment by timhemel** [comment]
The unfold behaviour was required because of a different interaction problem, which I have described here: http://www.securesoftware.nl/2016/09/15/gremlin-repeat-emit-select-strangeness/

Using identity instead of unfold does not fix that particular problem.

As far as the problem you're describing, a `sideEffect{ println it.dump()}` also changes the behaviour, which makes it a bit hard to debug.

If you manually expand the steps, you will get something like:
```
.until(...).repeat(__.in(AST_EDGE)).emit().repeat(out(AST_EDGE)
```

My guess is that it is not clear to which repetition the `emit` step belongs and a different traversal is constructed than you (and I) expected. The `emit` step is associated with the first repetition and not the second. Inserting almost any step will take away the confusion.

Good finding, it means that adding the identity step before a repetition would be another best practice.

**Comment by vlad902** [comment]
Yeah I considered submitting a patch that does just that, but am worried that any reasonable graph query optimization would just throw away that step and potentially lead to the same behavior. The TP3 docs mention an IdentityRemovalStrategy, but explain() doesn't seem to work for me so I can't confirm whether that traversal strategy is built-in and whether it succeeds with/despite it.

**Comment by timhemel** [comment]
Wrote a test and fixed the functions, they have been added to my outstanding pull request.

I don't think the optimization will get in the way, as I expect that the parsing will be done before any optimizations. `until` and `emit` are step modifiers, not actual steps, so they would be associated with the repetitions first, and then the traversal would be optimized. At least, that is how I would do it.

**Comment by vlad902** [comment]
Looks like octopus is using TP 3.0.1-incubating so it's quite old. Trying to fast forward it to 3.2.1 breaks octopus, but I'd be curious to see if that resolves it. If not, the explain() step should at least make it a little easier to understand how gremlin is interpreting it and what a correct solution is.

**Comment by vlad902** [comment]
Last release of TitanDB still references 3.0.1 and that might be why. Looks like TitanDB is EOL too so that's probably not going to change without a change of backend DBs :(
